pandas/core/util/_dispatching.py

```diff
@@ -0,0 +1,186 @@
+            'application': {'aggregate', 'transform'}
+            'cython_nm': ...  # Name of the Cython function to call
+            'extra_kwargs': {...}  # Extra kwargs to pass to Cython
+            'type_blacklist': [...]  # Dtypes for which func should raise
+            'result_type': ...  # dtype of result from Cython
+            'conversion_in': ...  # dtype or callable for conversion pre-Cython
+            'needs_values': ...  # Whether the obj values should pass to Cython
+            'needs_mask': ...   # Whether a mask of NA values should be passed
+            'conversion_out': ...  # dtype or callable for conv post-Cython
+        }
+        """
+        return {
+            'any': {
+                'application': 'aggregate',
+                'cython_nm' : 'group_any_all',
+                'extra_kwargs': {'val_test': 'any'},
+                'type_blacklist': [],
+                'result_type': np.uint8,
+                'conversion_in': self._any_all_convertor,
+                'needs_values': True,
+                'needs_mask': True,
+                'conversion_out': np.bool
+            },
+            'all': {
+                'application': 'aggregate',
+                'cython_nm' : 'group_any_all',
+                'extra_kwargs': {'val_test': 'all'},
+                'type_blacklist': [],
+                'result_type': np.uint8,
+                'conversion_in': self._any_all_convertor,
+                'needs_values': True,
+                'needs_mask': True,
+                'conversion_out': np.bool
+            }
+        }
+
+    @property
+    def application_type(self):
+        return self.func_metadata[self.func_nm]['application']
+
+    def _any_all_convertor(self, vals):
+        """
+        Converts objects to appropriate type for any/all calculations.
+        """
+        try:
+            vals = vals.astype(np.bool)
+        except ValueError:  # for objects
+            vals = np.array([bool(x) for x in vals])
+
+        return vals.view(np.uint8)
+
+    def _validate_types(self):
+        """
+        Validate that the types of the `grp_by` object.
+
+        Raises
+        ------
+        ``TypeError`` if the `grp_by` dtypes are not valid for `func_nm`.
+        """
+        if self.obj.values.dtype in self.func_metadata[
+                self.func_nm]['type_blacklist']:
+            raise TypeError("'{}' cannot be applied to a dtype of {}".format(
+                self.func_nm, self.obj.values.dtype))
+
+    def _get_result(self, **kwargs):
+        """
+        Fetch the result from the Cython layer.
+
+        Parameters
+        ----------
+        kwargs
+            Extra arguments to bind to the `func_nm` Cython signature.
+
+        Resolve function name in case of templating use.
+        """
+        # Since this func is called in a loop, the below might be better
+        # served outside of the loop and passed in?
+        labels, _, ngroups = self.groupby.grouper.group_info
+
```

Comment:
see above these should be subclasses

Reply:
I would make the things to dispatch just methods. Otherwise this has to have a lot of magic. Its much simpler to just call a method.

Reply:
Struggling to conceptualize this - so do you think it would be best to have a `BaseDispatcher` and then have the various groupings of applications as the subclasses (so here `AnyAllDispatcher`)?

Reply:
see my comments above 

first order of business is to move groupby to a sub module 
so can easily create more modules